### DISCUSSION TITLE
自定义控件有没有办法使用 formily 的组件？

### DISCUSSION THREAD

**Discussion by Zhang-theSenile** [alibaba/formily#2053] [Q&A 2.x]
我想做一个自定义控件（提供 value、onChange接口的组件），想在其内使用 ArrayCard、ArrayItems 之类的 formily 组件，有办法实现吗？

看「实现自定义组件」文档提到的「Hooks API」、「connect/mapProps/mapReadPretty API」也没有相关链接和示例。
看源码则因水平时间有限深感成本过高😅

**Comment by janryWang** [comment]
https://github.com/alibaba/formily/blob/formily_next/designable/setters/src/components/ValidatorSetter/index.tsx 这个组件就是使用了ArrayItems

**Reply by Zhang-theSenile** [reply]
尝试了一下，无报错也不显示控件，不知道问题是出在哪部分，可以帮忙看下吗？

custom-control.component.tsx
```tsx
import { ArrayItems } from '@formily/antdv2';  // @formily/antd@2.0.0-beta.22
import {
  ISchema,
  observer,
  Schema,
  SchemaContext,
} from '@formily/react';   // 2.0.0-beta.22


const schema: ISchema = {
  type: 'array',
  'x-component': 'ArrayItems',
  'x-decorator': 'FormItem',
  title: '对象数组',
  items: {
    type: 'object',
    properties: {
      space: {
        type: 'void',
        'x-component': 'Space',
        properties: {
          sort: {
            type: 'void',
            'x-decorator': 'FormItem',
            'x-component': 'ArrayItems.SortHandle',
          },
          date: {
            type: 'string',
            title: '日期',
            'x-decorator': 'FormItem',
            'x-component': 'DatePicker.RangePicker',
            'x-component-props': {
              style: {
                width: 160,
              },
            },
          },
          input: {
            type: 'string',
            title: '输入框',
            'x-decorator': 'FormItem',
            'x-component': 'Input',
          },
          select: {
            type: 'string',
            title: '下拉框',
            enum: [
              { label: '选项1', value: 1 },
              { label: '选项2', value: 2 },
            ],
            'x-decorator': 'FormItem',
            'x-component': 'Select',
            'x-component-props': {
              style: {
                width: 160,
              },
            },
          },
          remove: {
            type: 'void',
            'x-decorator': 'FormItem',
            'x-component': 'ArrayItems.Remove',
          },
        },
      },
    },
  },
  properties: {
    add: {
      type: 'void',
      title: '添加条目',
      'x-component': 'ArrayItems.Addition',
    },
  },
};

export const CustomArrayItems = observer(
  (props) => {
    return (
      <SchemaContext.Provider value={new Schema(schema)}>
        <ArrayItems />
      </SchemaContext.Provider>
    );
  },
);
```

my-form.tsx
```tsx
import { FormItem, Input, Radio } from '@formily/antdv2';  // @formily/antd@2.0.0-beta.22
import { createSchemaField } from '@formily/react';  // 2.0.0-beta.22
import { CustomArrayItems } from './custom-control.component'

const SchemaField = createSchemaField({
  components: {
    FormItem,
    Input,
    CustomArrayItems,
  },
});

function MyForm(props) {
  return (
    <SchemaField>
      <SchemaField.String
        required
        title="名称"
        name="name"
        x-decorator="FormItem"
        x-component="Input"
      />
      <SchemaField.Array
        title="自定义ArrayItems"
        name="rule"
        x-decorator="FormItem"  // 是否添加这行都不显示
        x-component="CustomArrayItems"
      />
    </SchemaField>
  );
}

```

**Reply by Zhang-theSenile** [reply]
摸索了一番。。似乎得在 my-form.tsx 的 `createSchemaForm()` 内提供 custom-control.component.tsx 所用的全部组件，比如 `Space` `Select` 之类的。。

有办法实现 调用方 无需关注 自定义控件内部依赖的组件吗？

**Reply by Zhang-theSenile** [reply]
又摸索了一番。。在 custom-control.component.tsx 再提供一下 context 就可以了，用的是 `SchemaOptionsContext`，components写法与 `createSchemaField()` 相同：
```tsx
<SchemaOptionsContext.Provider value={{ components: { FormItem, Input, ... } }}>
    <SchemaContext.Provider value={new Schema(schema)}>
        <ArrayItems />
    </SchemaContext.Provider>
</SchemaOptionsContext.Provider>
```
这些都是靠看源码接口试出来的😅   文档的话，好几个文档翻遍了也没找到解法

不过归根结底，依旧没搞懂为什么要传个 `<ArrayItems />` 作为 `<SchemaContext>` 的 children（看源码似乎是因为 ArrayItems 内部取得了 schema，据此做进一步渲染）
为什么 您给的 [ValidatorSetter](https://github.com/alibaba/formily/blob/formily_next/designable/setters/src/components/ValidatorSetter/index.tsx#L165) 不需要 provide components
为什么直接写 `<RecursionField schema={new Schema(schema)} />` 不行

**Reply by janryWang** [reply]
RecursionField可以的，只是在这个场景不行，这个场景比较复杂，它自身和内部都共用了一个key，x-validator，如果用RecursionField就会出现死循环渲染，所以才使用这个方案，有点偏hack了，我给的例子不太好。。。正常还是推荐RecursionField

**Reply by Zhang-theSenile** [reply]
可以问一下为什么您给的  [ValidatorSetter](https://github.com/alibaba/formily/blob/formily_next/designable/setters/src/components/ValidatorSetter/index.tsx#L165) 不需要 provide 内部用到的 component 吗？
我在唯一用到 ValidatorSetter 的 [DesignableField](https://github.com/alibaba/formily/blob/formily_next/designable/next/src/components/DesignableField/index.tsx#L194) 也没找到 component 的 provider

**Reply by janryWang** [reply]
不需要，它会通过context读取上下文中的组件集合

**Reply by Zhang-theSenile** [reply]
> 摸索了一番。。似乎得在 my-form.tsx 的 `createSchemaForm()` 内提供 custom-control.component.tsx 所用的全部组件，比如 `Space` `Select` 之类的。。
> 
> 有办法实现 调用方 无需关注 自定义控件内部依赖的组件吗？

不是很明白，意思就是像我最开始说的这样，需要在调用方提供全部的组件吗？

比如我的 form 只用了 Input 和 MyCustomField
而 MyCustomField 内部使用了 Select、Radio 等等
那么我在 form 里 createSchemaField 时就必须把 Select 和 Radio 也带上吗？

**Reply by janryWang** [reply]
嗯，必须提前传入所有依赖组件，渲染的时候才能指定

**Reply by Zhang-theSenile** [reply]
感谢您的回复

**Reply by 1793523411** [reply]
链接过期了，请问现在最新的formily里面可以 在自定义控件里使用 formily 的组件吗？ @janryWang
